app/src/main/java/com/duckduckgo/app/fire/CookieRemover.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2020 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.fire
+
+import android.database.DatabaseErrorHandler
+import android.database.sqlite.SQLiteDatabase
+import android.webkit.CookieManager
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.app.global.exception.UncaughtExceptionRepository
+import com.duckduckgo.app.global.exception.UncaughtExceptionSource
+import com.duckduckgo.app.statistics.pixels.Pixel
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+interface CookieRemover {
+    suspend fun removeCookies(): Boolean
+}
+
+class CookieManagerRemover(private val cookieManager: CookieManager) : CookieRemover {
+    override suspend fun removeCookies(): Boolean {
+        suspendCoroutine<Unit> { continuation ->
+            cookieManager.removeAllCookies {
+                Timber.v("All cookies removed; restoring DDG cookies")
+                continuation.resume(Unit)
+            }
+        }
+        return true
+    }
+}
+
+class SQLCookieRemover(
+    private val webViewDatabaseLocator: DatabaseLocator,
+    private val getCookieHostsToPreserve: GetCookieHostsToPreserve,
+    private val pixel: Pixel,
+    private val uncaughtExceptionRepository: UncaughtExceptionRepository,
+    private val dispatcherProvider: DispatcherProvider
+) : CookieRemover {
+
+    override suspend fun removeCookies(): Boolean {
+        return withContext(dispatcherProvider.io()) {
+            val databasePath: String = webViewDatabaseLocator.getDatabasePath()
+            if (databasePath.isNotEmpty()) {
+                val excludedHosts = getCookieHostsToPreserve()
+                return@withContext removeCookies(databasePath, excludedHosts)
+            } else {
+                pixel.fire(Pixel.PixelName.COOKIE_DATABASE_NOT_FOUND)
```

Comment:
There might be a race condition here between the pixel trying to fire and the app process trying to restart. I think you might see some cases where pixels are lost because of this.

In other places where we want a pixel near to the app restart, we store the pixel offline and send them when the app next syncs.

Reply:
Oh, thanks for pointing out that issue. Interesting 😅

I have a question: I'm sending this pixel to detect something unexpected happening related to the database and to detect it somehow from our side (e.g: grafana + alert). I was also using `recordUncaughtException` to receive the stack trace and to have some more info about the crash, but...

Will it be enough if we only rely on using:
`recordUncaughtException(exception, UncaughtExceptionSource.COOKIE_DATABASE)`?
I think exception will be stored on the database and we will send the pixel when app next syncs, but I prefer to double-check with you.

Reply:
In this context though, you aren't crashing the app if there's a problem. The app might be about to restart, but it's not because of an uncaught exception, so I don't think it makes sense to try use uncaught exception handler here to get the pixel data. Maybe that makes sense in some of the other cases though?

Reply:
For the record. Here I'm not crashing the app, but neither in other parts related to the WebView database (if an error occurs, I capture the exception and fallback to removing everything via cookie manager). The difference with this case vs the others, it's that here I'm not capturing any exception whereas in the others I'm capturing the exception triggered by an external component and sending it back to us to know what's the issue with more detail.

Let's talk about this offline so we can discuss the correct approach for this. I would like to find a simple workaround instead of building a component to store pixels and trigger them when the app restarts only for this case.

Reply:
We talked about this offline: We are going to remove any usage of `recordUncaughtException`(1) here. Instead we will use `fire.pixel` (maybe losing some of them due to app restarting) to send a sneak peek of the stack trace as a parameter and use `OfflinePixelSender` to keep a real counter of how many times an issue happens.

(1) Reason we are not using `recordUncaughtException` is that we handle those exceptions. Responsibility for that class is to register any exception that makes the app to crash. We will introduce a new way of registering handled exceptions or illegal states but not in the scope of this PR.